pkg/server/api/localauthority/v1/service.go

```diff
@@ -0,0 +1,278 @@
+}
+
+func (s *Service) GetX509AuthorityState(ctx context.Context, _ *localauthorityv1.GetX509AuthorityStateRequest) (*localauthorityv1.GetX509AuthorityStateResponse, error) {
+	log := rpccontext.Logger(ctx)
+
+	var states []*localauthorityv1.AuthorityState
+	current := s.ca.GetCurrentX509CASlot()
+	if !current.IsEmpty() {
+		state, err := slotToProto(current, localauthorityv1.AuthorityState_ACTIVE)
+		if err != nil {
+			return nil, api.MakeErr(log, codes.Internal, "failed to get current slot", err)
+		}
+		states = append(states, state)
+	}
+
+	next := s.ca.GetNextX509CASlot()
+	// when next has a key indicates that it was initialized
+	if next.GetPublicKey() != nil {
+		status := localauthorityv1.AuthorityState_PREPARED
+		// CA is removed from slot on rotation
+		if next.IsEmpty() {
+			status = localauthorityv1.AuthorityState_OLD
+		}
+
+		state, err := slotToProto(next, status)
+		if err != nil {
+			return nil, api.MakeErr(log, codes.Internal, "failed to get next slot", err)
+		}
+		states = append(states, state)
+	}
+
+	rpccontext.AuditRPC(ctx)
+
+	return &localauthorityv1.GetX509AuthorityStateResponse{
+		States: states,
+	}, nil
+}
+
+func (s *Service) PrepareX509Authority(ctx context.Context, req *localauthorityv1.PrepareX509AuthorityRequest) (*localauthorityv1.PrepareX509AuthorityResponse, error) {
+	log := rpccontext.Logger(ctx)
+
+	if err := s.ca.PrepareX509CA(ctx); err != nil {
+		return nil, api.MakeErr(log, codes.Internal, "failed to prepare X.509 authority", err)
+	}
+
+	slot := s.ca.GetNextX509CASlot()
+	// Prepare is going to use current slot when it is empty
+	if slot.IsEmpty() {
+		slot = s.ca.GetCurrentX509CASlot()
+	}
+
+	authorityState, err := slotToProto(slot, localauthorityv1.AuthorityState_PREPARED)
+	if err != nil {
+		return nil, api.MakeErr(log, codes.Internal, "failed to create response", err)
+	}
+
+	rpccontext.AuditRPC(ctx)
+
+	return &localauthorityv1.PrepareX509AuthorityResponse{
+		PreparedAuthority: authorityState,
+	}, nil
+}
+
+func (s *Service) ActivateX509Authority(ctx context.Context, req *localauthorityv1.ActivateX509AuthorityRequest) (*localauthorityv1.ActivateX509AuthorityResponse, error) {
+	log := rpccontext.Logger(ctx)
+
+	if s.ca.GetNextX509CASlot().IsEmpty() {
+		return nil, api.MakeErr(log, codes.Internal, "no prepared authority found", nil)
+	}
+
+	// Move next into current and reset next to clean CA
+	s.ca.RotateX509CA()
+
+	current := s.ca.GetCurrentX509CASlot()
+
+	state, err := slotToProto(current, localauthorityv1.AuthorityState_ACTIVE)
+	if err != nil {
+		return nil, api.MakeErr(log, codes.Internal, "failed to parse current slot", err)
+	}
+
+	rpccontext.AuditRPC(ctx)
+
+	return &localauthorityv1.ActivateX509AuthorityResponse{
+		ActivatedAuthority: state,
+	}, nil
+}
+
+func (s *Service) TaintX509Authority(ctx context.Context, req *localauthorityv1.TaintX509AuthorityRequest) (*localauthorityv1.TaintX509AuthorityResponse, error) {
+	parseRequest := func() logrus.Fields {
```

Comment:
@MarcosDY What do you think about deanonymizing this function? it is duplicated in the method below